### PR TITLE
Fix reading/writing from urllib.request objects

### PR DESCRIPTION
**Pull request by mdboom**
Addresses matplotlib-devel thread "Possible bug in matplotlib versions 1.4 and higher with basemap arcgisimage imagery"

### CODE REVIEW

File: lib/matplotlib/tests/test_image.py [hunk 1]
```diff
@@ -517,6 +517,13 @@ def test_minimized_rasterized():
                 assert False
 
 
+@cleanup
```
Comment (jenshnielsen) [inline lib/matplotlib/tests/test_image.py]: doesn't this need a @attr('network') decorator too?

Comment (mdboom) [inline lib/matplotlib/tests/test_image.py]: Indeed.  I forgot we had that.

File: src/_png.cpp [hunk 2]
```diff
@@ -206,7 +206,17 @@ static PyObject *Py_write_png(PyObject *self, PyObject *args, PyObject *kwds)
             goto exit;
         }
         buff.cursor = 0;
-    } else if ((fp = mpl_PyFile_Dup(py_file, (char *)"wb", &offset))) {
+    } else {
+        #if PY3K
```
Comment (WeatherGod) [inline src/_png.cpp]: Just remembered seeing an article recently imploring that we stop writing if-statements like this that would possibly break in py4.x. The logic really should be checking whether or not it is py2.x, as py2.x should be considered the exception to the rule.

Comment (mdboom) [inline src/_png.cpp]: Nope -- this won't break.  Here's the definition of it in mpl_utils.h:

```c++
#if PY_MAJOR_VERSION >= 3
#define PY3K 1
#define Py_TPFLAGS_HAVE_NEWBUFFER 0
#else
#define PY3K 0
#endif
```

I take `PY3K` to mean "next-gen Python" as it was called before Python 3.0 was even released.

**Comment by WeatherGod** [comment]
This all looks good to me. The appveyor failures are the usual ones. Thanks for the quick turnaround! I'll also backport to v1.5.x branch.

**Comment by WeatherGod** [comment]
Ok, actually, the backport isn't totally clean. I wonder why?

**Comment by WeatherGod** [comment]
Looks like we never backported this code portion:
```
     if (filein == Py_None) {
         buff.size = width * height * 4 + 1024;
         buff.str = PyBytes_FromStringAndSize(NULL, buff.size);
         if (buff.str == NULL) {
             goto exit;
         }
         buff.cursor = 0;
     } else {
```
Was that intentional?

**Comment by jenshnielsen** [comment]
Does git blame tell us where it is coming from?

**Comment by WeatherGod** [comment]
looks like b3e9443b

**Comment by jenshnielsen** [comment]
So that traces it to https://github.com/matplotlib/matplotlib/pull/5389 which is mainly a performance improvement and thus not backported. Not sure how to best handle that

**Comment by WeatherGod** [comment]
Looking over it, I wonder if that was the commit that broke things? So,
maybe the bug isn't on v1.5.x?

**Comment by jenshnielsen** [comment]
According to the original issue mail the bug is on 1.4.0 and up so I guess that

**Comment by WeatherGod** [comment]
hmmm, that's right. How about I take a crack at the backport and come up with a PR for it so that we can double-check that it makes sense?

**Comment by jenshnielsen** [comment]
@WeatherGod :+1:

**Comment by mdboom** [comment]
> Looks like we never backported this code portion.  Was this intentional?

Yes -- it's an optimization for the nbagg backend.  It doesn't make sense to backport that unless we backport the related changes to nbagg (which I wouldn't recommend).

**Comment by WeatherGod** [comment]
#5927 is my attempt at backporting to v1.5.x

**Pull request merged**